lib/icinga/apiactions.cpp

```diff
@@ -488,7 +488,11 @@ Dictionary::Ptr ApiActions::RemoveDowntime(const ConfigObject::Ptr& object,
 				downtime->SetRemovedBy(author);
 			}
 
-			Downtime::RemoveDowntime(downtime->GetName(), true);
+			try {
+				Downtime::RemoveDowntime(downtime->GetName(), true);
+			} catch (std::runtime_error& error) {
+				return ApiActions::CreateResult(400, error.what());
```

Comment:
This could give a strange response code if there's either something else already raising a `std::runtime_error` or if something like this would get introduced in the future. Using a new custom exception type would be much nicer.

Reply:
What should the new exception type be called?  how about ``used_object_error``?

Reply:
If you don't have any plan to re-use this somewhere else, I'd go for a more descriptive name, like for example `invalid_downtime_removal_error` (quite long unfortunately).

Another idea would be to make it something that can clearly mapped to a 400 response code and also reused in other places (possibly one could even use `std::invalid_argument`, better than `std::runtime_error` I guess).

Reply:
But if you look [here](https://github.com/Icinga/icinga2/blob/bugfix/remove-downtime-returns-wrong-status-7408/lib/icinga/downtime.cpp#L340) even in the whole ``dowtime.cpp`` file, a ``std::runtime_error`` is thrown if the downtime object cannot be deleted or created and not ``std::invalid_argument``.

Reply:
Does it have to be the same exception everywhere? Like in the end, it would be nice to return a status code that matches what happened. So if the failure is related to user input, return some 4xx status (what you tried to delete does not exist, you don't have the permission to delete this, you tried to delete a downtime created from a scheduled one, ...), and if something went wrong within Icinga, return 5xx, most likely 500 internal server error (out of memory, filesystem error when writing the change to disk, ...). When using `std::runtime_error` everywhere, this is hard to achieve, note that for example `std::system_error` extends from `std::runtime_error` and that is an exception that should almost certainly result in a 500 Internal Server Error. I know, Icinga isn't perfect in this regard at the moment, but it would still be nice to get this right for new code.